pkg/cloud/utils.go

```diff
@@ -128,9 +139,25 @@ func (r *ResourceID) String() string {
 //	[https://www.googleapis.com/compute/<ver>]/projects/<proj>/global/<res>/<name>
 //	[https://www.googleapis.com/compute/<ver>]/projects/<proj>/regions/<region>/<res>/<name>
 //	[https://www.googleapis.com/compute/<ver>]/projects/<proj>/zones/<zone>/<res>/<name>
+//
+// Note that ParseResourceURL can't round trip partial paths that do not
+// include an API Group.
 func ParseResourceURL(url string) (*ResourceID, error) {
 	errNotValid := fmt.Errorf("%q is not a valid resource URL", url)
 
+	matches := apiGroupRegex.FindStringSubmatch(url)
```

Comment:
Can we make this more precise instead of finding arbitrary submatches? Ideally, we strip off the constant prefix and match exact from start to end of string.

Reply:
Unfortunately this function supports a _very_ broad range of inputs and there is no shared prefix we could rely on here:

https://github.com/GoogleCloudPlatform/k8s-cloud-provider/blob/ade9202145be71e04bd11fd952a7829d1fab7c5d/pkg/cloud/utils.go#L119-L130

Reply:
This is particularly relevant because we allow the `googleapis.com` domain to be overridden.

Reply:
That doesn't mean we can't make the regex more specific...

Reply:
As far as I can tell, we can't do anything to predict what comes before API Group because it's configurable via `SetAPIDomain`. We do know that `/projects` will _always_ be after it, and usually a `<version>` path segment will come between API Group and `/projects`. That's what the RegEx is matching for. Any suggestions for how we could make this more specific?

Reply:
Updated regex to only match known set of versions and added more test cases, let me know if you'd rather revert to something a bit looser here.

Reply:
We should put validation in SetDomain to restrict the domain of that.